### PR TITLE
[FLINK-3931] Implement Transport Encryption (SSL/TLS)

### PR DESCRIPTION
**Pull request by skrishnappa**
Enabled SSL support for the different network transport within flink.
This feature is documented as task T3 in the "Secure Data Access in Flink" design doc - https://docs.google.com/document/d/1-GQB6uVOyoaXGwtqwqLV8BHDxWiMO2WnVzBoJ8oPaAs

This commit addresses subtasks - FLINK-4324, FLINK-4325, FLINK-4404 and FLINK-4405
For details about how to use this feature, refer to the new setup guide at security-ssl.md and the new config values added to config.md

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	 * Starts the BLOB server.
+	 */
+	@BeforeClass
+	public static void startServer() {
+		try {
+			Configuration config = new Configuration();
+			config.setBoolean(ConfigConstants.SECURITY_SSL_ENABLED, true);
+			config.setString(ConfigConstants.SECURITY_SSL_KEYSTORE, "src/test/resources/local127.keystore");
+			config.setString(ConfigConstants.SECURITY_SSL_KEYSTORE_PASSWORD, "password");
+			config.setString(ConfigConstants.SECURITY_SSL_KEY_PASSWORD, "password");
```
Comment (mxm) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java]: Ah, it is enabled by default when the global SSL flag is set. Can we fail the test if SSL is not enabled?

**Comment by mxm** [comment]
CC @StephanEwen You might want to take a look as well.

File: docs/setup/config.md [hunk 2]
```diff
@@ -140,6 +140,8 @@ will be used under the directory specified by jobmanager.web.tmpdir.
 
 - `blob.server.port`: Port definition for the blob server (serving user jar's) on the Taskmanagers. By default the port is set to 0, which means that the operating system is picking an ephemeral port. Flink also accepts a list of ports ("50100,50101"), ranges ("50100-50200") or a combination of both. It is recommended to set a range of ports to avoid collisions when multiple JobManagers are running on the same machine.
 
+- `blob.service.ssl.enabled`: Flag to enable ssl for the blob client/server communication. This is applicable only when the global ssl flag security.ssl.enabled is set to true (DEFAULT: true).
```
Comment (skrishnappa) [inline docs/setup/config.md]: Setting security.ssl.enabled to true will enable ssl for all communication. The other flags are required only for selectively disabling ssl (they are set to true by default). The reasons I chose to provide the extra flags are the following
* web frontend - this is useful if the admin chooses to encrypt only external traffic
* taskmanager data trasnfer - enabling ssl here might have significant performance impact and the admin might choose to encrypt only management and control traffic

I am slightly biased towards keeping these extra config (based on past experiences) and wasn't sure what the community would prefer. I am fine removing these and doing SSL - all or nothing, please let me know.

File: docs/setup/config.md [hunk 3]
```diff
@@ -181,6 +183,7 @@ The following parameters configure Flink's JobManager and TaskManagers.
 - `taskmanager.hostname`: The hostname of the network interface that the TaskManager binds to. By default, the TaskManager searches for network interfaces that can connect to the JobManager and other TaskManagers. This option can be used to define a hostname if that strategy fails for some reason. Because different TaskManagers need different values for this option, it usually is specified in an additional non-shared TaskManager-specific config file.
 - `taskmanager.rpc.port`: The task manager's IPC port (DEFAULT: **0**, which lets the OS choose a free port).
 - `taskmanager.data.port`: The task manager's port used for data exchange operations (DEFAULT: **0**, which lets the OS choose a free port).
+- `taskmanager.data.ssl.enabled`: Enable SSL support for the taskmanager data transport. This is applicable only when the global ssl flag security.ssl.enabled is set to true (DEFAULT: **true**)
```
Comment (skrishnappa) [inline docs/setup/config.md]: please see previous comment related to this

File: docs/setup/config.md [hunk 4]
```diff
@@ -222,6 +238,7 @@ The following parameters configure Flink's JobManager and TaskManagers.
 - `jobmanager.web.backpressure.refresh-interval`: Time after which available stats are deprecated and need to be refreshed (DEFAULT: `60000`, 1 min).
 - `jobmanager.web.backpressure.num-samples`: Number of stack trace samples to take to determine back pressure (DEFAULT: `100`).
 - `jobmanager.web.backpressure.delay-between-samples`: Delay between stack trace samples to determine back pressure (DEFAULT: `50`, 50 ms).
+- `jobmanager.web.ssl.enabled`: Enable https access to the web frontend. This is applicable only when the global ssl flag security.ssl.enabled is set to true (DEFAULT: `true`).
```
Comment (skrishnappa) [inline docs/setup/config.md]: please see previous comment related to this

File: docs/setup/security-ssl.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+* Restart the flink components to enable SSL for all of flink's internal communication
+* Verify by accessing the jobmanager's UI using https url. The task manager's path in the UI should show akka.ssl.tcp:// as the protocol
+* The blob server and task manager's data communication can be verified from the log files
+
+## YARN Deployment
+The keystores and truststore can be deployed in a YARN setup in multiple ways depending on the cluster setup. Following are 2 ways to achieve this
+
+### 1. Deploy keystores before starting the YARN session
+The keystores and truststore should be generated and deployed on all nodes in the YARN setup where flink components can potentially be executed. The same flink config file from the flink YARN client is used for all the flink components running in the YARN cluster. Therefore we need to ensure the keystore is deployed and accessible using the same filepath in all the YARN nodes.
```
Comment (skrishnappa) [inline docs/setup/security-ssl.md]: Yes. This is specifically useful in cases where companies have their own PKI infrastructure and would want to re-use the certificates and its location for different applications

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 6]
```diff
@@ -157,6 +157,9 @@
 	 */
 	public static final String BLOB_SERVER_PORT = "blob.server.port";
 
+	/** Flag to override ssl support for the blob service transport */
+	public static final String BLOB_SERVICE_SSL_ENABLED = "blob.service.ssl.enabled";
+
```
Comment (skrishnappa) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: please see previous comment related to this

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 7]
```diff
@@ -178,6 +181,11 @@
 	public static final String TASK_MANAGER_DATA_PORT_KEY = "taskmanager.data.port";
 
 	/**
+	 * Config parameter to override SSL support for taskmanager's data transport
+	 */
+	public static final String TASK_MANAGER_DATA_SSL_ENABLED = "taskmanager.data.ssl.enabled";
```
Comment (skrishnappa) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: please see previous comment related to this

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java (hunk 1 above)
Comment (skrishnappa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java]: Sure, I'll add another test case for that

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java [hunk 8]
```diff
@@ -304,8 +314,15 @@ private void respondAsLeader(ChannelHandlerContext ctx, HttpRequest request, Str
...
-		ctx.write(new DefaultFileRegion(raf.getChannel(), 0, fileLength), ctx.newProgressivePromise());
-		ChannelFuture lastContentFuture = ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
+		ChannelFuture lastContentFuture;
+		if (ctx.pipeline().get(SslHandler.class) == null) {
+			ctx.write(new DefaultFileRegion(raf.getChannel(), 0, fileLength), ctx.newProgressivePromise());
+			lastContentFuture = ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
+		} else {
+			lastContentFuture = ctx.writeAndFlush(new HttpChunkedInput(new ChunkedFile(raf, 0, fileLength, 8192)),
+				ctx.newProgressivePromise());
+			// HttpChunkedInput will write the end marker (LastHttpContent) for us.
```
Comment (skrishnappa) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: This is based on the Netty example code - https://netty.io/4.0/xref/io/netty/example/http/file/HttpStaticFileServerHandler.html

The initial flink code somehow did not include these checks. Cannot do zero copy file transfer with SSL enabled and the existing stream handling code hangs without any data transfer. Hence had to introduce this.

File: docs/setup/config.md (hunk 2 above)
Comment (mxm) [inline docs/setup/config.md]: Fair point, we might want to keep them for a flexible setup. As long as the default is to encrypt all channels when the global flag is turned on, they don't really get in the way of the user.

Comment (skrishnappa) [inline docs/setup/config.md]: Yeah, thats how it is currently implemented. So I'll leave these configuration as is.

**Comment by mxm** [comment]
@skrishnappa I would like to merge this. Could you add the test case to ensure encryption is turned on and rebase the PR?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java (hunk 1 above)
Comment (skrishnappa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSslTest.java]: Added few more testcases for failure scenarios

**Comment by skrishnappa** [comment]
Have included few more test cases to handle ssl failure scenarios.
Also rebased the PR to the latest master

**Comment by mxm** [comment]
Thanks for the update @skrishnappa. The test changes look good. Merging the PR to master.

**Comment by mxm** [comment]
Thanks for your contribution @skrishnappa!

**Comment by skrishnappa** [comment]
Thanks @mxm for following this up. Cheers

**Pull request closed without merging**